Trustee unified CLI

spotlesstofu opened the pull request:
Hey! I've been trying to address #529 and bootstrap a CLI for Trustee.

### Aliases and plugins
I'm supporting aliases and plugins.
Aliases are when you link to the same executable with different names (à la busybox).
Plugins allow third parties to create a `trustee-something` executable and run it with
```
trustee something do-my-third-party-thing
```

### Structure / design
The design direction I took is to unify all the command lines in a single one.
I'm opening this pull as a draft while it's in early stages. Below there is a graphic of the structure I'm thinking about.

![trustee-new-architecture drawio](https://github.com/user-attachments/assets/32f82786-fcc5-4532-9240-1290e215e456)

### Next steps

I'll be updating this pull in the next days. This is what I have in mind right now:

- Implement more commands.
- Make the `run` command spawn all the components separately.
- Allow to `run` without a config file, using defaults + CLI arguments.
- Use functions that already exist in the codebase for the `keygen` command.

Harsh feedback is welcome 😁

fitzthum commented:
It's great to see this PR. A few high-level comments 

First, some simple things. I'd probably find a different location for the CLI code; maybe in the tools repo? Also, I wonder if we need the keygen command to be exposed on its own. Maybe we could just generate the key automatically if it doesn't already exist.

A bigger question is whether the CLI should spawn any other processes. I lean towards not doing that. I like how you are starting the KBS a a crate like we do in the CI. This is suited to the local Trustee approach (where you have a short-lived trustee running locally that can attest something else like another Trustee in an enclave). We should also think about if the CLI can hook into docker compose or the operator to deploy a long-lived Trustee.

spotlesstofu commented:
Pushed new commits.

Partially addressed @fitzthum comments:
- Moved the module to `tools/`.
- Now generating the admin auth key automatically if it's not set.

I also added a self-signed certificate generation and set a default policy, see commits.

I think exposing the keygen command is still useful. It enables to integrate it in a workflow that we are not yet supporting automatically. For example one may want to use it to create a secret on Kubernetes?

To do:
- Do not spawn processes.
- Use the trustee_home directory as a base for other file paths, like the attestation service work_dir.

These two I'd like to address in this pull request. Then I'll open new pulls to further extend the features of the tool.

pmores commented on tools/trustee/src/cli.rs:
```diff
@@ -50,6 +83,9 @@ pub async fn cli_default() -> Result<(), Error> {
             trustee_keygen(&out).unwrap();
         }
+        Commands::Run { config_file } => {
+            trustee_run(config_file, &trustee_home_dir).await.unwrap();
```
Do we want to go with `unwrap()` here (and other commands as well)?  With `unwrap()` I might get
```
thread 'main' panicked at tools/trustee/src/cli.rs:126:18:
called `Result::unwrap()` on an `Err` value: refusing to overwrite file: "/home/pvl/.trustee/https_key.pem"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
whereas if I replace the `unwrap()` with just a quick fix like this:
```
.map_err(|e| Error::raw(clap::error::ErrorKind::InvalidValue, format!("{}\n", e.to_string())))?
```
the same error is reported as
```
error: refusing to overwrite file: "/home/pvl/.trustee/https_key.pem"
```
which seems much clearer.  Granted, both contain the important information but frankly I initially didn't even notice it in the unwrap message. :-)

pmores replied:
A broader question might be how we want to handle errors.  The fix I proposed above assumes we want to return `clap::Error` from `cli_default()` - but do we?  The only reason seems to be the `Cli::try_parse()?` call at the top of the function, command handlers apparently use `anyhow`.

Maybe we'd prefer to convert the `try_parse()` error to `anyhow` and then `cli_default()` returns anyhow errors, making it possible for command handlers to be called simply like
```
trustee_run(config_file, &trustee_home_dir).await?;
```
?

pmores commented:
Also, unless I revert the initial commit (but keep `derive(Default)` on `KbsConfig`) `KbsConfig` seems to be unable to read the sample config file in `kbs/config/kbs-config.toml` anymore due to
```
invalid config: missing field `sockets`
```
If `sockets` is supplied it complains about the next config member.  Basically all configuration items seem to have become mandatory.

spotlesstofu commented:
> ```
> invalid config: missing field `sockets`
> ```

So weird. If I add a `derive(Default)` it works fine.

It's like `try_deserialize` is ignoring the `impl Default` but not the `derive(Default)`.

spotlesstofu commented:
> It's like `try_deserialize` is ignoring the `impl Default` but not the `derive(Default)`.

My bad. I confused with `#[serde(default)]`. I'll push a fix.

spotlesstofu commented:
Comments addressed and rebased on main. I'd appreciate a review.

fitzthum commented on tools/trustee/src/main.rs:
```diff
@@ -0,0 +1,38 @@
+
+    #[cfg(feature = "aliases")]
+    match aliases::match_alias() {
+        Ok(_) => return,
```
Maybe I don't understand the aliases feature at all but why would we return here?

fitzthum commented on tools/trustee/src/main.rs:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    #[cfg(feature = "plugins")]
+    match std::env::args().nth(1) {
```
I don't really like reading the args manually like this. You are using clap in cli.rs but not here.

fitzthum commented on tools/trustee/src/cli.rs:
```diff
@@ -0,0 +1,169 @@
+            let out = out.unwrap_or_else(|| trustee_home_dir.join("key"));
+            trustee_keygen(&out)?;
+        }
+        Commands::Run {
```
May want to make this enum variant more explicit in terms of what it is running.

fitzthum reviewed (commented):
A few comments. 

I still don't really understand the purpose of plugins or aliases. In the code, the relationship between `main.rs` and `cli.rs` is a little confusing. It seems like the CLI parsing should potentially be the top level.

fitzthum commented on tools/trustee/src/cli.rs:
```diff
@@ -0,0 +1,169 @@
+    Ok(config)
+}
+
+async fn trustee_run(config: KbsConfig) -> Result<()> {
```
This is a bit different because it's running Trustee as a crate directly from the CLI process. This is intended to run locally potentially to bootstrap another Trustee running inside an enclave.

fitzthum commented:
Yeah, I would just drop the alias support for now to make this PR as simple as possible. Keep the code and we can add it back in if someone comes with the requirement.

For plugins I would prefer that we access any sub-programs via crates rather than as binaries. I don't have a strong opinion about what goes into this PR for plugins.

Otherwise seems good.

spotlesstofu commented on kbs/src/attestation/intel_trust_authority/mod.rs:
```diff
@@ -84,6 +84,7 @@ pub struct IntelTrustAuthorityConfig {
     pub api_key: String,
     pub certs_file: String,
     pub allow_unmatched_policy: Option<bool>,
+    #[serde(default)]
```
Indeed. I removed it.

spotlesstofu replied:
No. Removing it makes the test fail. I put it back.
```
invalid config: missing field `policy_ids`
```

pmores replied:
Which test is that?  `cargo test` is failing for me in `test tests::parse_data_json_binding::case_3` and
`token::simple::tests::test_issue_simple_ephemeral_key`, and that's _regardless_ of this hunk...

pmores commented on tools/trustee/src/cli.rs:
```diff
@@ -0,0 +1,171 @@
+    };
+
+    config.policy_engine.policy_path =
+        replace_base_dir(config.policy_engine.policy_path.as_path(), trustee_home_dir);
```
Is it guaranteed that the policy path - and other paths passed to `replace_base_dir()` later in this function - will have the base path that `replace_base_dir()` assumes?

Or is it the intended semantics that we redirect specifically and only `/opt/confidential-containers/*` paths to `trustee_home_dir` but leave any other paths alone?  If so I think it would warrant a short explanation in a code comment.

pmores commented:
Cool, you even managed to handle the default value of `--home` with the `default_value` annotation - I didn't think this was even possible for options of a type which is not `Display`. :-)

This seems much easier to make sense of, at least to me.  Now, I'm still not clear about the purpose of the `keygen` command.  I'm aware that this has been raised and discussed before but... looking at the code it appears that `keygen` could be used to generate the `auth` key pair though not the HTTPS key pair which uses a different algorithm.  However, the `run` command will generate it too, if it's missing.  What is the advantage of generating the `auth` key pair ahead of time using `keygen`?

spotlesstofu commented:
Yes, `keygen` generates the auth key.

pmores commented:
Right.  I understand the idea of shielding users from raw `openssl` commands and I know it can be easy to misuse them inadvertently.  But isn't that goal achieved already by having the `run` command generate `auth` if necessary?

Or, again, what is then the advantage of using `keygen` to generate `auth` explicitly?

spotlesstofu commented:
I think of these use cases for the `keygen` command:
1. Generate the key pair for a remote deployment (Kubernetes). Because of the security model for private-public key pairs, a remote deployment can't auto-generate the key. The admin can generate the key with `keygen` and provide the public key to the remote deployment. This is a stand-alone usage of `keygen` as a utility and doesn't involve the `run` command.
2. The user wants to make a custom config file and reference the key from there. Then the config file can be passed to the `run` command.

pmores commented:
Quite a bunch of paths are retargeted to trustee home dir in `run::get_config()` but not the resource repository path.  Even if I give `--home` my resources are stored under `/opt/confidential-containers/kbs/repository` (the  value of `dir_path` of the `LocalFs` plugin in my KBS config file).  Is that an omission, or intentional?

pmores reviewed (approved):
lgtm thanks @spotlesstofu for your patience! :-)

fitzthum reviewed (approved):
LGTM. 

There are some things to tweak but I think this is a good start.

mkulke commented on tools/trustee/src/main.rs:
```diff
@@ -0,0 +1,106 @@
+/// Error if the file already exists.
+fn write_pem(private_path: &Path, public_path: &Path, private_key: &PKey<Private>) -> Result<()> {
+    // Write the private key to a file private to the owner.
+    {
```
why the block scope? is it to flush f?
